### ISSUE TITLE
Convert to/from php-ast nodes

### ISSUE THREAD

**Issue by jesseschalken** [nikic/PHP-Parser#279]
I have a lot of code built on top of this project's APIs, but the php-ast extension seems likely to be faster at parsing than PHP-Parser as it exposes PHP's own parser written in C.

Would it be possible to support conversion between the two AST formats? The conversion would be lossy since PHP-Parser exposes more information (comments not attached to declarations, start/end file offset etc), but that information isn't necessary for many purposes.

**Comment by nikic** [comment]
Should be possible and I'd love to see this :) I don't think I have the time right now to implement this myself.

**Comment by TysonAndre** [comment]
I started working on converting \PHPParser\Node to \ast\Node, but right now, most node types are stubs/TODOs. See https://github.com/TysonAndre/php-parser-to-php-ast

- The final goal is to create equivalent objects to php-ast, for the test files used by php-ast, when converting from \PHPParser\Node to \ast\Node (With the exception of line numbers of some node types)
- The test suite is set up. 22 of 28 tests are failing.
- No plans to convert \ast\Node to \PHPParser\Node in that repo.
- Will update if this is completed

**Comment by nikic** [comment]
Nice work! For validation I would recommend looping over all php-src phpt tests and comparing the parser output. The test suite in this extension is rather incomplete, as it mainly tests the parts that are rewritten in some way.

**Comment by TysonAndre** [comment]
I'm finished implementing https://github.com/TysonAndre/php-parser-to-php-ast (One way conversion to php-ast)

I looped over php-ast's tests and added a few overlooked node types(goto, label, __halt_compiler() (I didn't implement the offset lookup yet, but that's doable (e.g. with php tokenizer or searching for that line)).
I also fixed some incorrecly set flags.

- Start line numbers are slightly different in the output for statement lists, etc.
- There's two modes of execution taking advantage of PhpParser's error tolerance: Omit statements with errors, or try to add placeholders. (e.g. convert `$a = $this->` to `$a = $this->__INCOMPLETE_PROPERTY__`)
   These aren't tested yet.
